proposals/NNNN-generalize-async-sequence.md

```diff
@@ -0,0 +1,311 @@
+    }
+  }
+}
+```
+
+Note that the default implementation of `nextElement` necessarily violates `Sendable` checking in order to pass `self` from a possibly-isolated context to a `nonisolated` one. Though this is generally unsafe, this is how calls to `next()` behave today, so existing conformances will maintain the behavior they already have. Implementing `nextElement()` directly will eliminate the unsafety.
+
+To enable conformances of `AsyncIteratorProtocol` to only implement `nextElement()`, a default implementation is also provided for `next()`:
+
+```swift
+extension AsyncIteratorProtocol {
+  @available(SwiftStdlib 5.11, *)
+  public mutating func next() async throws(Failure) -> Element? {
+    // Callers to `next()` will always run `nextElement()` on the generic executor.
+    try await nextElement(nil)
+  }
+}
+```
+
+Both function requirements of `AsyncIteratorProtocol` have default implementations that are written in terms of each other, meaning that it is a programmer error to implement neither of them. Types that are available prior to the Swift 5.11 standard library must provide an implementation of `next()`, because the default implementation is only available with the Swift 5.11 standard library.
+
+To avoid silently allowing conformances that implement neither requirement, and to facilitate the transition of conformances from `next()` to `nextElement()`, we add a new availability rule where the witness checker diagnoses a protocol conformance that uses an deprecated, obsoleted, or unavailable default witness implementation. Deprecated implementations will produce a warning, while obsoleted and unavailable implementations will produce an error.
+
+Because the default implementation of `nextElement()` is deprecated, conformances that do not provide a direct implementation will produce a warning. This is desirable because the default implementation of `nextElement()` violates `Sendable` checking, so while it's necessary for source compatibilty, it's important to aggressively suggest that conforming types implement the new method.
+
+### Associated type inference for `AsyncIteratorProtocol` conformances
+
+When an `AsyncIteratorProtocol`-conforming type provides a `nextElement` function, the `Failure` type is inferred based on whether (and what) `nextElement` throws using the rules described in [SE-0413](/proposals/0413-typed-throws.md).
+
+If the `AsyncIteratorProtocol`-conforming type uses the default implementation of `nextElement`, then the `Failure` associated type is inferred from the `next` function instead based on the following rules:
+
+* If `next()` throws nothing, `Failure` is inferred to `Never`.
+* If `next()` throws, `Failure` is inferred to `any Error`.
+* If `next()` rethrows, `Failure` is inferred to `T.Failure`, where `T` is the first type parameter with a conformance to either `AsyncSequence` or `AsyncIteratorProtocol`. If there are multiple such requirements, take the `errorUnion` of them all.
```

Comment:
Also here—if `rethrows` for `next()` isn't a 'real' thing, do we need to spell it out explicitly in this list?